<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Проект-&quot;Приложение-для-тех,-кто-любит-читать&quot;-SQL" data-toc-modified-id="Проект-&quot;Приложение-для-тех,-кто-любит-читать&quot;-SQL-1">Проект "Приложение для тех, кто любит читать" SQL</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1.0.1">Описание проекта</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.0.2">Описание данных</a></span></li></ul></li><li><span><a href="#Шаг-1.-Импорт-библиотек-и-настройка-коннектора" data-toc-modified-id="Шаг-1.-Импорт-библиотек-и-настройка-коннектора-1.1">Шаг 1. Импорт библиотек и настройка коннектора</a></span></li><li><span><a href="#Шаг-2.-Познакомимся-с-данными-таблиц" data-toc-modified-id="Шаг-2.-Познакомимся-с-данными-таблиц-1.2">Шаг 2. Познакомимся с данными таблиц</a></span><ul class="toc-item"><li><span><a href="#2.1-Таблица-с-книгами" data-toc-modified-id="2.1-Таблица-с-книгами-1.2.1">2.1 Таблица с книгами</a></span></li><li><span><a href="#2.2.-Таблица-с-авторами" data-toc-modified-id="2.2.-Таблица-с-авторами-1.2.2">2.2. Таблица с авторами</a></span></li><li><span><a href="#2.3.-Таблица-publishers-с-издательствами" data-toc-modified-id="2.3.-Таблица-publishers-с-издательствами-1.2.3">2.3. Таблица publishers с издательствами</a></span></li><li><span><a href="#2.4.-Таблица-ratings-с-рейтингами-книг" data-toc-modified-id="2.4.-Таблица-ratings-с-рейтингами-книг-1.2.4">2.4. Таблица ratings с рейтингами книг</a></span></li><li><span><a href="#2.5.-Таблица-reviews-с-обзорами-книг" data-toc-modified-id="2.5.-Таблица-reviews-с-обзорами-книг-1.2.5">2.5. Таблица reviews с обзорами книг</a></span></li></ul></li><li><span><a href="#Шаг-3.-Исследование" data-toc-modified-id="Шаг-3.-Исследование-1.3">Шаг 3. Исследование</a></span><ul class="toc-item"><li><span><a href="#3.1.-Посчитаем,-сколько-книг-вышло-после-1-января-2000-года." data-toc-modified-id="3.1.-Посчитаем,-сколько-книг-вышло-после-1-января-2000-года.-1.3.1">3.1. Посчитаем, сколько книг вышло после 1 января 2000 года.</a></span></li><li><span><a href="#3.2.-Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку." data-toc-modified-id="3.2.-Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку.-1.3.2">3.2. Для каждой книги посчитаем количество обзоров и среднюю оценку.</a></span></li><li><span><a href="#3.3.-Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключим-из-анализа-брошюры;" data-toc-modified-id="3.3.-Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключим-из-анализа-брошюры;-1.3.3">3.3. Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;</a></span></li><li><span><a href="#3.4.-Определим-автора-с-самой-высокой-средней-оценкой-книг-—-будем-учитывать-только-книги-с-50-и-более-оценками;" data-toc-modified-id="3.4.-Определим-автора-с-самой-высокой-средней-оценкой-книг-—-будем-учитывать-только-книги-с-50-и-более-оценками;-1.3.4">3.4. Определим автора с самой высокой средней оценкой книг — будем учитывать только книги с 50 и более оценками;</a></span></li><li><span><a href="#3.5.-Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="3.5.-Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-1.3.5">3.5. Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li></ul></li></ul></div>

# Проект "Приложение для тех, кто любит читать" SQL

### Описание проекта

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

**Задача**
Проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных 

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

##  Шаг 1. Импорт библиотек и настройка коннектора 

In [1]:
# импортируем библиотеки
import pandas as pd

from sqlalchemy import create_engine


In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], db_config['pwd'], db_config['host'], db_config['port'],
db_config['db'])
#сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Шаг 2. Познакомимся с данными таблиц

### 2.1 Таблица с книгами

In [3]:
query = '''
        Select *
        From books
        Limit 1
       '''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id         title  num_pages publication_date  publisher_id
0        1        546  'Salem's Lot        594       2005-11-01            93

### 2.2. Таблица с авторами

In [4]:
query = '''
        Select * 
        From authors
        Limit 1
       '''
pd.io.sql.read_sql(query, con = engine)

author_id      author
0          1  A.S. Byatt

### 2.3. Таблица publishers с издательствами

In [5]:
query = '''
        Select * 
        From publishers
        Limit 1
       '''
pd.io.sql.read_sql(query, con = engine)

publisher_id publisher
0             1       Ace

### 2.4. Таблица ratings с рейтингами книг

In [6]:
query = '''
        Select * 
        From ratings
        Limit 1
       '''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id    username  rating
0          1        1  ryanfranco       4

### 2.5. Таблица reviews с обзорами книг

In [7]:
query = '''
        Select * 
        From reviews
        Limit 1
       '''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id      username  \
0          1        1  brandtandrea   

                                                text  
0  Mention society tell send professor analysis. ...

## Шаг 3. Исследование

### 3.1. Посчитаем, сколько книг вышло после 1 января 2000 года.

In [9]:
query = '''
        Select COUNT(book_id)
        From books
        WHERE publication_date > '2000-01-01'
       '''
pd.io.sql.read_sql(query, con = engine)

count
0    819

**После 1 января 2000 года было выпущено 819 книг из 1000 представленных книг**

### 3.2. Для каждой книги посчитаем количество обзоров и среднюю оценку.


In [10]:
query = '''
        SELECT
        book_id, title Название_книги, COUNT(review_id) Число_обзоров, ROUND(AVG(rating),2)_Средний_рейтинг
        FROM 
        books LEFT OUTER JOIN reviews using(book_id)
        LEFT OUTER JOIN ratings using(book_id)
        GROUP BY 1
        ORDER BY 1
       '''
pd.io.sql.read_sql(query, con = engine)

book_id                                     Название_книги  \
0          1                                       'Salem's Lot   
1          2                 1 000 Places to See Before You Die   
2          3  13 Little Blue Envelopes (Little Blue Envelope...   
3          4  1491: New Revelations of the Americas Before C...   
4          5                                               1776   
..       ...                                                ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997                        Xenocide (Ender's Saga  #3)   
997      998                                    Year of Wonders   
998      999                        You Suck (A Love Story  #2)   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...   

     Число_обзоров  _Средний_рейтинг  
0                6              3.67  
1                2              2.50  
2                9              4.67  
3                4              4.50  
4               24              4.00  
..             ...               ...  
995              9              3.67  
996             15              3.40  
997             20              3.20  
998              4              4.50  
999             24              3.83  

[1000 rows x 4 columns]

### 3.3. Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;

In [11]:
query = '''
        SELECT publisher AS Издательство
        FROM
        (SELECT publisher,COUNT(book_id)
        FROM publishers
        INNER JOIN books using(publisher_id)
        WHERE num_pages > 50
        GROUP BY 1
        ORDER BY 2 DESC
        LIMIT 1) AS g
       '''
pd.io.sql.read_sql(query, con = engine)

Издательство
0  Penguin Books

**Издательство `Penguin Books` выпустило наибольшее количество книг толше 50 страниц (42 книги)**

### 3.4. Определим автора с самой высокой средней оценкой книг — будем учитывать только книги с 50 и более оценками;

In [12]:
query = '''
        
        SELECT author Автор
        FROM
        (WITH t1 AS(SELECT book_id,ROUND(AVG(rating),2) AS avg
        From ratings 
        GROUP BY 1
        HAVING COUNT(rating) > 50),
        t2 AS(SELECT book_id, author
                    FROM books INNER JOIN authors using(author_id))
        SELECT author, AVG(avg)
        FROM t1 LEFT OUTER JOIN t2 using(book_id)
        GROUP BY 1
        ORDER BY 2 DESC
        LIMIT 1) AS d
        
        '''
pd.io.sql.read_sql(query, con = engine)

Автор
0  J.K. Rowling/Mary GrandPré

**`J.K. Rowling/Mary GrandPré`  - это автор с самой высокой средней оценкой книг. Скорее всего, именно произведения этого автора Вам могут понравиться**

### 3.5. Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [13]:
query = '''
  WITH t1 AS(SELECT username, COUNT(review_id) AS Количество_обзоров
            FROM reviews
            WHERE username IN(SELECT username
            FROM ratings
            GROUP BY 1
            HAVING COUNT(rating) > 50)
            GROUP BY 1)
  SELECT ROUND(AVG(Количество_обзоров),0) AS Среднее_количество_обзоров
         FROM t1
        '''
pd.io.sql.read_sql(query, con = engine)

Среднее_количество_обзоров
0                        24.0

**Итак, всего 6 пользователей поставили больше 50 оценок и мы видим их среднее количество обзоров, которое составило 24 обзора. Наиболее читающие люди? Или люди, любящие выссказать свое мнение? В любом случае, именно благодаря им, вы сможете составить предварительное мнение о книге и решить, стоит ее читать или нет.**

**Итак, у нас было 5 таблиц с разной информацией.**
 Нам удалось объединить информацию воедино. Мы проделали большую работу, чтобы получить ответы на важные вопросы и облегчить читателям поиски лучшей книги для себя.
 Новые таблицы помогут нам понять:
 1. Востребовано ли приложение? 
 2. Много ли читающих активных пользователей?
 2. Какую книгу выбрать для чтения с помощью рейтинга и числа обзоров?
 3. Посмотреть средний рейтинг книги, автора.
 4. Прежде чем начать читать книгу, познакомиться со мнением других читателей, что поможет принять окончательное решение.
